File: python/day2/main.py
import re

INPUT_FILE = "input.sf"


def part1(lines):
    result = 0
    constraints = {"red": 12, "green": 13, "blue": 14}
    for id, line in enumerate(lines, start=1):
        if all(
            map(
                lambda elem: int(elem[0]) <= constraints[elem[1]],
                re.findall(r"(\d*) (red|green|blue)", line),
            )
        ):
            result += id
    return result


def part2(lines):
    result = 0
    for line in lines:
        record = {"red": 0, "green": 0, "blue": 0}
        for amount, color in re.findall(r"(\d*) (red|green|blue)", line):
            if int(amount) > record[color]:
                record[color] = int(amount)
        result += record["red"] * record["green"] * record["blue"]
    return result


f = open(INPUT_FILE)
file_lines = f.readlines()

print("part1:", part1(file_lines))
print("part2:", part2(file_lines))

f.close()

File: perl/day2/main.pl
use warnings;
use strict;

my $input_file = "input.sf";

open( my $input_fh, "<", $input_file ) || die "could not open file";

my @lines = <$input_fh>;

my %max_cubes = ( "red" => 12, "green" => 13, "blue" => 14 );
my ( $part1, $part2 ) = 0;

foreach my $line (@lines) {
    my @cubes       = reverse $line =~ /(\d+) (blue|red|green)/g;
    my $line_number = ( $line =~ /Game (\d+)/ )[0];

    my $all    = 1;
    my %record = ( "red" => 0, "green" => 0, "blue" => 0 );

    while ( my ( $i, $j ) = splice( @cubes, 0, 2 ) ) {
        if ( $max_cubes{$i} < $j ) { $all = 0; }
        if ( $record{$i} < $j ) { $record{$i} = $j; }
    }

    $part2 += $record{"red"} * $record{"green"} * $record{"blue"};

    if ($all) { $part1 += $line_number; }
}

print "part1: $part1\npart2: $part2\n";

close($input_fh);
